README.md
# Async Weather SDK

[![PyPI version](https://img.shields.io/pypi/v/async-weather-sdk?logo=python&logoColor=white)](https://badge.fury.io/py/async-weather-sdk)
[![GitHub Workflow Status for tests](https://img.shields.io/github/workflow/status/decentfox/async-weather-sdk/test?logo=github&logoColor=white)](https://github.com/decentfox/async-weather-sdk/actions?query=workflow%3Atest)
[![Codacy Badge](https://img.shields.io/codacy/coverage/f548667427c24fc394204b440166c26d?logo=Codacy)](https://www.codacy.com/gh/decentfox/async-weather-sdk?utm_source=github.com&utm_medium=referral&utm_content=decentfox/async-weather-sdk&utm_campaign=Badge_Coverage)

Async weather API wrapper for fetching weather and forecast data

## Core Dependencies

**Asyncio:** a library to write concurrent code using the async/await syntax.

**Aiohttp:** an asynchronous HTTP Client/Server for asyncio and Python.

## Install

```bash
pip install async-weather-sdk

OR

poetry add async-weather-sdk
```

## Usage

### QQ Weather API SDK

Get current weather/forecast data by province and city.

```python
from async_weather_sdk.qq import QQWeather

weather = QQWeather()

await weather.fetch_current_weather('北京市', '北京市')
await weather.fetch_weather_forecast('北京市', '北京市', 3)
```

Query current weather/forecast data with tencent map api key.

```python
from async_weather_sdk.qq import query_current_weather, query_weather_forecast

await query_current_weather('API_KEY', '北京市')
await query_weather_forecast('API_KEY', '39.90469,116.40717')
```


HISTORY.rst
=======
History
=======

0.1.1 (2020-06-02)
^^^^^^^^^^^^^^^^^^

* First release on PyPI.


tests/test_qq.py
import sys

import asyncio
import aiohttp
import pytest

from async_weather_sdk.qq import query_current_weather, query_weather_forecast
from async_weather_sdk.qq import QQMap, QQWeather

pytestmark = pytest.mark.asyncio


async def test_query_current_weather(aresponses, qq_forecast_resp):
    with pytest.raises(ValueError, match="Please provide tencent map api key"):
        await query_current_weather("", "北京")

    with pytest.raises(ValueError, match="Empty query"):
        await query_current_weather("API_KEY", "")

    aresponses.add(
        "apis.map.qq.com",
        "/ws/location/v1/ip",
        "GET",
        response={
            "status": 0,
            "message": "query ok",
            "result": {
                "ip": "61.135.17.68",
                "location": {"lat": 39.90469, "lng": 116.40717},
                "ad_info": {
                    "nation": "中国",
                    "province": "北京市",
                    "city": "北京市",
                    "district": "",
                    "adcode": 110000,
                },
            },
        },
    )
    aresponses.add(
        "wis.qq.com", "/weather/common", "GET", response=qq_forecast_resp,
    )

    res = await query_current_weather("API_KEY", "61.135.17.68")
    assert "observe" in res
    assert res["observe"] == {
        "degree": "29",
        "humidity": "30",
        "precipitation": "0.0",
        "pressure": "998",
        "update_time": "202006011323",
        "weather": "晴",
        "weather_code": "00",
        "weather_short": "晴",
        "wind_direction": "5",
        "wind_power": "2",
    }
    assert res["rise"] == {
        "sunrise": "04:47",
        "sunset": "19:36",
        "time": "20200601",
    }
    assert res["location"] == {
        "nation": "中国",
        "province": "北京市",
        "city": "北京市",
        "district": "",
        "adcode": 110000,
    }


async def test_query_weather_forecast(aresponses, qq_forecast_resp):
    with pytest.raises(ValueError, match="Please provide tencent map api key"):
        await query_weather_forecast("", "北京")

    with pytest.raises(ValueError, match="Empty query"):
        await query_weather_forecast("API_KEY", "")

    with pytest.raises(ValueError, match="Invalid forecast days"):
        await query_weather_forecast("API_KEY", "北京", 10)

    aresponses.add(
        "apis.map.qq.com",
        "/ws/location/v1/ip",
        "GET",
        response={
            "status": 0,
            "message": "query ok",
            "result": {
                "ip": "61.135.17.68",
                "location": {"lat": 39.90469, "lng": 116.40717},
                "ad_info": {
                    "nation": "中国",
                    "province": "北京市",
                    "city": "北京市",
                    "district": "",
                    "adcode": 110000,
                },
            },
        },
    )
    aresponses.add(
        "wis.qq.com", "/weather/common", "GET", response=qq_forecast_resp,
    )

    res = await query_weather_forecast("API_KEY", "61.135.17.68", 2)
    assert res == {
        "rise": [
            {"sunrise": "04:47", "sunset": "19:36", "time": "20200601"},
            {"sunrise": "04:47", "sunset": "19:37", "time": "20200602"},
        ],
        "forecast": [
            {
                "day_weather": "小雨",
                "day_weather_code": "07",
                "day_weather_short": "小雨",
                "day_wind_direction": "西北风",
                "day_wind_direction_code": "7",
                "day_wind_power": "4",
                "day_wind_power_code": "1",
                "max_degree": "26",
                "min_degree": "14",
                "night_weather": "晴",
                "night_weather_code": "00",
                "night_weather_short": "晴",
                "night_wind_direction": "西风",
                "night_wind_direction_code": "6",
                "night_wind_power": "3",
                "night_wind_power_code": "0",
                "time": "2020-05-31",
            },
            {
                "day_weather": "雷阵雨",
                "day_weather_code": "04",
                "day_weather_short": "雷阵雨",
                "day_wind_direction": "西南风",
                "day_wind_direction_code": "5",
                "day_wind_power": "5",
                "day_wind_power_code": "2",
                "max_degree": "30",
                "min_degree": "16",
                "night_weather": "多云",
                "night_weather_code": "01",
                "night_weather_short": "多云",
                "night_wind_direction": "西南风",
                "night_wind_direction_code": "5",
                "night_wind_power": "3",
                "night_wind_power_code": "0",
                "time": "2020-06-01",
            },
            {
                "day_weather": "晴",
                "day_weather_code": "00",
                "day_weather_short": "晴",
                "day_wind_direction": "南风",
                "day_wind_direction_code": "4",
                "day_wind_power": "5",
                "day_wind_power_code": "2",
                "max_degree": "33",
                "min_degree": "20",
                "night_weather": "晴",
                "night_weather_code": "00",
                "night_weather_short": "晴",
                "night_wind_direction": "西南风",
                "night_wind_direction_code": "5",
                "night_wind_power": "4",
                "night_wind_power_code": "1",
                "time": "2020-06-02",
            },
        ],
        "location": {
            "nation": "中国",
            "province": "北京市",
            "city": "北京市",
            "district": "",
            "adcode": 110000,
        },
    }


async def test_qq_weather_sdk_fetch_weather(aresponses):
    aresponses.add(
        "wis.qq.com",
        "/weather/common",
        "GET",
        response={
            "message": "OK",
            "status": 200,
            "data": {
                "observe": {
                    "degree": "29",
                    "humidity": "30",
                    "precipitation": "0.0",
                    "pressure": "998",
                    "update_time": "202006011323",
                    "weather": "晴",
                    "weather_code": "00",
                    "weather_short": "晴",
                    "wind_direction": "5",
                    "wind_power": "2",
                },
            },
        },
    )

    async with aiohttp.ClientSession() as session:
        qq_weather = QQWeather(session=session)

        res = await qq_weather.fetch_weather("北京", "北京", "observe")
        assert "observe" in res

    aresponses.add(
        "wis.qq.com",
        "/weather/common",
        "GET",
        response={"status": 311, "message": "key格式错误"},
    )

    async with aiohttp.ClientSession() as session:
        qq_weather = QQWeather(session=session)

        res = await qq_weather.fetch_weather("北京", "北京", "observe")
        assert res == {}


async def test_qq_weather_sdk_fetch_current_weather(
    aresponses, qq_forecast_resp
):
    aresponses.add(
        "wis.qq.com", "/weather/common", "GET", response=qq_forecast_resp,
    )
    async with aiohttp.ClientSession() as session:
        qq_weather = QQWeather(session=session)
        res = await qq_weather.fetch_current_weather("北京", "北京")
        assert "observe" in res
        assert "sunrise" in res["rise"]


async def test_qq_weather_sdk_fetch_weather_forecast(
    aresponses, qq_forecast_resp
):
    aresponses.add(
        "wis.qq.com", "/weather/common", "GET", response=qq_forecast_resp,
    )

    async with aiohttp.ClientSession() as session:
        qq_weather = QQWeather(session=session)
        res = await qq_weather.fetch_weather_forecast("北京", "北京", 3)

        assert "forecast" in res
        assert len(res["forecast"]) == 4

        assert "rise" in res
        assert len(res["rise"]) == 3

    aresponses.add(
        "wis.qq.com", "/weather/common", "GET", response=qq_forecast_resp,
    )

    async with aiohttp.ClientSession() as session:
        qq_weather = QQWeather(session=session)
        res = await qq_weather.fetch_weather_forecast("北京", "北京", 1)

        assert "forecast" in res
        assert len(res["forecast"]) == 25

        assert "rise" in res
        assert len(res["rise"]) == 1


async def test_qq_map_wrong_api_key(aresponses):
    aresponses.add(
        "apis.map.qq.com",
        "/ws/location/v1/ip",
        "GET",
        response={"status": 311, "message": "key格式错误"},
    )

    async with aiohttp.ClientSession() as session:
        qq_map = QQMap("API_KEY", session=session)
        res = await qq_map.location_lookup_by_ip("61.135.17.68")
        assert res == {}


async def test_qq_map_location_lookup_by_ip(aresponses):
    aresponses.add(
        "apis.map.qq.com",
        "/ws/location/v1/ip",
        "GET",
        response={
            "status": 0,
            "message": "query ok",
            "result": {
                "ip": "61.135.17.68",
                "location": {"lat": 39.90469, "lng": 116.40717},
                "ad_info": {
                    "nation": "中国",
                    "province": "北京市",
                    "city": "北京市",
                    "district": "",
                    "adcode": 110000,
                },
            },
        },
    )
    async with aiohttp.ClientSession() as session:
        qq_map = QQMap("API_KEY", session=session)

        res = await qq_map.location_lookup_by_ip("61.135.17.68")
        assert res == {
            "nation": "中国",
            "province": "北京市",
            "city": "北京市",
            "district": "",
            "adcode": 110000,
        }


async def test_qq_map_location_lookup_by_coordinates(aresponses):
    aresponses.add(
        "apis.map.qq.com",
        "/ws/geocoder/v1",
        "GET",
        response={
            "status": 0,
            "message": "query ok",
            "result": {
                "location": {"lat": 39.90469, "lng": 116.40717},
                "address": "北京市东城区正义路2号",
                "ad_info": {
                    "nation_code": "156",
                    "adcode": "110101",
                    "city_code": "156110000",
                    "name": "中国,北京市,北京市,东城区",
                    "location": {"lat": 39.916668, "lng": 116.434578},
                    "nation": "中国",
                    "province": "北京市",
                    "city": "北京市",
                    "district": "东城区",
                },
            },
        },
    )
    async with aiohttp.ClientSession() as session:
        qq_map = QQMap("API_KEY", session=session)

        res = await qq_map.location_lookup_by_coordinates("39.90469,116.40717")
        assert res == {
            "nation_code": "156",
            "adcode": "110101",
            "city_code": "156110000",
            "name": "中国,北京市,北京市,东城区",
            "location": {"lat": 39.916668, "lng": 116.434578},
            "nation": "中国",
            "province": "北京市",
            "city": "北京市",
            "district": "东城区",
        }

    aresponses.add(
        "apis.map.qq.com",
        "/ws/geocoder/v1",
        "GET",
        response={"status": 400, "message": "query failed",},
    )
    async with aiohttp.ClientSession() as session:
        qq_map = QQMap("API_KEY", session=session)

        res = await qq_map.location_lookup_by_coordinates("39.90469,116.40717")
        assert res == {}


async def test_qq_map_location_lookup_by_keyword(aresponses):
    aresponses.add(
        "apis.map.qq.com",
        "/ws/district/v1/search",
        "GET",
        response={
            "status": 0,
            "message": "query ok",
            "result": [
                [
                    {
                        "id": "110000",
                        "name": "北京",
                        "fullname": "北京市",
                        "pinyin": ["bei", "jing"],
                        "level": 1,
                        "location": {"lat": 39.90469, "lng": 116.40717},
                        "address": "北京",
                    },
                    {
                        "id": "230225580",
                        "fullname": "北京市双河农场",
                        "level": 4,
                        "location": {"lat": 47.866631, "lng": 123.753351},
                        "address": "黑龙江,齐齐哈尔,甘南县,北京市双河农场",
                    },
                ]
            ],
        },
    )
    aresponses.add(
        "apis.map.qq.com",
        "/ws/geocoder/v1",
        "GET",
        response={
            "status": 0,
            "message": "query ok",
            "result": {
                "location": {"lat": 39.90469, "lng": 116.40717},
                "address": "北京市东城区正义路2号",
                "ad_info": {
                    "nation_code": "156",
                    "adcode": "110101",
                    "city_code": "156110000",
                    "name": "中国,北京市,北京市,东城区",
                    "location": {"lat": 39.916668, "lng": 116.434578},
                    "nation": "中国",
                    "province": "北京市",
                    "city": "北京市",
                    "district": "东城区",
                },
            },
        },
    )
    async with aiohttp.ClientSession() as session:
        qq_map = QQMap("API_KEY", session=session)

        res = await qq_map.location_lookup_by_keyword("北京市")
        assert res == {
            "nation_code": "156",
            "adcode": "110101",
            "city_code": "156110000",
            "name": "中国,北京市,北京市,东城区",
            "location": {"lat": 39.916668, "lng": 116.434578},
            "nation": "中国",
            "province": "北京市",
            "city": "北京市",
            "district": "东城区",
        }

    aresponses.add(
        "apis.map.qq.com",
        "/ws/district/v1/search",
        "GET",
        response={"status": 0, "message": "query ok", "result": [],},
    )
    async with aiohttp.ClientSession() as session:
        qq_map = QQMap("API_KEY", session=session)

        res = await qq_map.location_lookup_by_keyword("北京市")
        assert res == {}

    aresponses.add(
        "apis.map.qq.com",
        "/ws/district/v1/search",
        "GET",
        response={"status": 400, "message": "query failed",},
    )
    async with aiohttp.ClientSession() as session:
        qq_map = QQMap("API_KEY", session=session)

        res = await qq_map.location_lookup_by_keyword("北京市")
        assert res == {}


@pytest.fixture()
def mock_location_lookup_by_ip(mocker):
    future = asyncio.Future()
    mocker.patch(
        "async_weather_sdk.qq.QQMap.location_lookup_by_ip", return_value=future
    )
    future.set_result("mock_location_lookup_by_ip")
    return future


@pytest.fixture()
def mock_location_lookup_by_coordinates(mocker):
    future = asyncio.Future()
    mocker.patch(
        "async_weather_sdk.qq.QQMap.location_lookup_by_coordinates",
        return_value=future,
    )
    future.set_result("location_lookup_by_coordinates")
    return future


@pytest.fixture()
def mock_location_lookup_by_keyword(mocker):
    future = asyncio.Future()
    mocker.patch(
        "async_weather_sdk.qq.QQMap.location_lookup_by_keyword",
        return_value=future,
    )
    future.set_result("location_lookup_by_keyword")
    return future


async def test_qq_map_location_lookup(
    mock_location_lookup_by_ip,
    mock_location_lookup_by_coordinates,
    mock_location_lookup_by_keyword,
):
    if sys.version_info >= (3, 8):
        return

    qq_map = QQMap("API_KEY")
    res = await qq_map.location_lookup("61.135.17.68")
    assert res == "mock_location_lookup_by_ip"

    res = await qq_map.location_lookup("39.90469,116.40717")
    assert res == "location_lookup_by_coordinates"

    res = await qq_map.location_lookup("北京市")
    assert res == "location_lookup_by_keyword"


async def test_qq_map_location_lookup_p38(mocker):
    if sys.version_info < (3, 8):
        return

    mocker.patch("async_weather_sdk.qq.QQMap.location_lookup_by_ip")
    mocker.patch("async_weather_sdk.qq.QQMap.location_lookup_by_coordinates")
    mocker.patch("async_weather_sdk.qq.QQMap.location_lookup_by_keyword")

    qq_map = QQMap("API_KEY")
    await qq_map.location_lookup("61.135.17.68")
    QQMap.location_lookup_by_ip.assert_called_once()

    await qq_map.location_lookup("39.90469,116.40717")
    QQMap.location_lookup_by_coordinates.assert_called_once()

    await qq_map.location_lookup("北京市")
    QQMap.location_lookup_by_keyword.assert_called_once()


tests/test_client.py
import aiohttp
import asyncio
import pytest
from aiohttp import web

from async_weather_sdk.base import BaseClient

pytestmark = pytest.mark.asyncio


async def test_client_instantiation():
    client = BaseClient("https://BASE_ENDPOINT/")

    assert client.session is None

    async with aiohttp.ClientSession() as session:
        client = BaseClient("https://BASE_ENDPOINT/", session=session)
        assert client.session is not None

        with pytest.raises(aiohttp.ClientConnectionError):
            await client.request("/API_ENDPOINT")

        assert not session.closed


def test_client_get_url():
    fake_client = BaseClient("https://BASE_ENDPOINT/")

    assert (
        fake_client._get_url("/API_ENDPOINT")
        == "https://BASE_ENDPOINT/API_ENDPOINT"
    )

    assert (
        fake_client._get_url("https://API_ENDPOINT") == "https://API_ENDPOINT"
    )


async def test_request_with_error_response(aresponses):
    aresponses.add(
        "BASE_ENDPOINT",
        "/v1",
        "GET",
        response=aresponses.Response(status=500, text="error"),
    )

    async with aiohttp.ClientSession() as session:
        client = BaseClient("https://BASE_ENDPOINT/", session=session)

        res = await client.request("/v1")
        assert res == "error"


async def test_request_with_success_response(aresponses):
    aresponses.add(
        "BASE_ENDPOINT",
        "/v1",
        "GET",
        response=aresponses.Response(
            status=200,
            text='{"status": 200}',
            headers={"CONTENT-TYPE": "application/json"},
        ),
    )

    async with aiohttp.ClientSession() as session:
        client = BaseClient("https://BASE_ENDPOINT/", session=session)

        res = await client.request("/v1")
        assert res == {"status": 200}


async def test_request_with_timeout(aresponses):
    async def response_handler(request):
        await asyncio.sleep(1)
        return aresponses.Response(text="ok")

    aresponses.add("BASE_ENDPOINT", "/v1", "GET", response_handler)
    with pytest.raises(web.HTTPBadRequest, match="HTTP Request Timeout"):
        async with aiohttp.ClientSession() as session:
            client = BaseClient("https://BASE_ENDPOINT/", session=session)

            await client.request("/v1", timeout=0.05)


async def test_session_closed_after_request(aresponses):
    aresponses.add(
        "BASE_ENDPOINT",
        "/v1",
        "GET",
        response=aresponses.Response(
            status=200,
            text='{"status": 200}',
            headers={"CONTENT-TYPE": "application/json"},
        ),
    )

    client = BaseClient("https://BASE_ENDPOINT/")
    assert client.session is None
    await client.request("/v1")


src/async_weather_sdk/base.py
import logging
from typing import Optional
from urllib.parse import urljoin

import aiohttp
import asyncio
from aiohttp import web


class BaseClient(object):
    endpoint = None

    def __init__(
        self,
        endpoint: str,
        session: Optional[aiohttp.ClientSession] = None,
        logger: Optional[logging.Logger] = None,
    ):
        """
        Implement client that performs weather API requests.

        :param endpoint: The base endpoint URL
        :param session: Optionally specify the aiohttp session
        :param logger: An optional logger
        """
        self.endpoint = endpoint or self.endpoint
        self.logger = logger or logging.getLogger(__name__)
        self.session = session

    def _get_url(self, url):
        if self.endpoint and not url.startswith(("http://", "https://")):
            return urljoin(self.endpoint, url)
        return url

    async def request(
        self, url: str, method: Optional[str] = "GET", **aio_kwargs
    ):
        self.logger.debug("Fetch data from %s, %s", url, aio_kwargs)
        req_url = self._get_url(url)
        session = self.session or aiohttp.ClientSession(raise_for_status=True)
        try:
            async with session.request(method, req_url, **aio_kwargs) as resp:
                if "json" in resp.headers.get("CONTENT-TYPE"):
                    res = await resp.json()
                else:
                    res = await resp.text()
                self.logger.debug("Data fetched %r", res)
                return res
        except asyncio.TimeoutError:
            raise web.HTTPBadRequest(reason="HTTP Request Timeout")
        except (
            aiohttp.ClientResponseError,
            aiohttp.ClientConnectionError,
        ) as e:
            self.logger.warning("Error when getting %s, %s", url, e)
            raise e
        finally:
            if not self.session and not session.closed:
                await session.close()


src/async_weather_sdk/qq.py
import re
import logging
from typing import Optional

import aiohttp

from .base import BaseClient

WEATHER_ENDPOINT = "https://wis.qq.com"
MAP_ENDPOINT = "https://apis.map.qq.com"

qq_logger = logging.getLogger(__name__)


class QQWeather(BaseClient):
    def __init__(
        self,
        session: Optional[aiohttp.ClientSession] = None,
        logger: Optional[logging.Logger] = None,
    ):
        """
        Implement QQ Weather client that performs QQ weather API requests.

        :param session: Optionally specify the aiohttp session
        :param logger: An optional logger
        """
        super().__init__(
            endpoint=WEATHER_ENDPOINT, session=session, logger=logger
        )

    async def fetch_weather(self, province: str, city: str, weather_type: str):
        """
        Fetch weather data from Tencent (QQ) Weather API.

        :param province: Province Name in Chinese, for example: 北京市
        :param city: City Name in Chinese, for example: 北京市
        :param weather_type: Available types:
            observe|forecast_1h|forecast_24h|index|alarm|limit|tips|rise|air

            observe - Return real-time weather.
            forecast_1h - Return weather forecast data split hourly
                          in next 24 hours.
            forecast_24h - Return weather forecast data split daily
                           in next 7 days.
            index - Return today's index data of living.
            alarm - Return real-time weather alarm data.
            limit - Return today's car limit data.
            tips - Return today's weather tips data.
            rise - Return sunrise and sunset data.
            air - Return real-time air quality data.
        :return: Weather API response data.
        """
        params = dict(
            source="pc",
            weather_type=weather_type or "",
            province=province or "",
            city=city or "",
        )
        res = await self.request("/weather/common", params=params)
        if res.get("status") == 200 and res.get("message") == "OK":
            return res["data"]
        return {}

    async def fetch_current_weather(self, province: str, city: str) -> dict:
        """
        Return current weather data.

        :param province: Province Name in Chinese, for example: 北京市
        :param city: City Name in Chinese, for example: 北京市
        :return: real-time weather data.
        """
        res = await self.fetch_weather(
            province, city, "observe|index|alarm|limit|tips|rise|air"
        )
        res.update(rise=res.get("rise", {}).get("0", {}),)
        return res

    async def fetch_weather_forecast(
        self, province: str, city: str, forecast_days: int = 7
    ) -> dict:
        """
        Return weather forecast data for up to 7 days into the future.

        :param province: Province Name in Chinese, for example: 北京市
        :param city: City Name in Chinese, for example: 北京市
        :param forecast_days: The number of days for which the API returns
                              forecast data (Default: 7 days).
                              If pass forecast_days is 1, it will return
                              weather data split hourly.
        :return: forecast weather data.
        """
        forecast_days = min((max(1, forecast_days), 7))
        weather_type = "forecast_24h|rise"
        if forecast_days == 1:
            weather_type = "forecast_1h|rise"
        res = await self.fetch_weather(province, city, weather_type)
        if forecast_days == 1:
            weather_data = sorted(
                res.get("forecast_1h", {}).values(),
                key=lambda item: item["update_time"],
            )
        else:
            weather_data = sorted(
                res.get("forecast_24h", {}).values(),
                key=lambda item: item["time"],
            )
        if forecast_days > 1:
            weather_data = weather_data[: forecast_days + 1]
        else:
            weather_data = weather_data[:25]

        rise_data = sorted(
            res.get("rise", {}).values(), key=lambda item: item["time"]
        )
        return dict(forecast=weather_data, rise=rise_data[:forecast_days],)


class QQMap(BaseClient):
    api_key = None

    def __init__(
        self,
        api_key: str,
        session: Optional[aiohttp.ClientSession] = None,
        logger: Optional[logging.Logger] = None,
    ):
        """
        Implement QQ Map client that performs QQ Map API requests.

        :param api_key: QQ Map WebService API key
        :param session: Optionally specify the aiohttp session
        :param logger: An optional logger
        """
        self.api_key = api_key
        super().__init__(endpoint=MAP_ENDPOINT, session=session, logger=logger)

    async def location_lookup_by_ip(self, ip: str):
        params = dict(ip=ip, key=self.api_key)
        res = await self.request("/ws/location/v1/ip", params=params)
        if res.get("status") != 0:
            self.logger.warning("Failed to query location by IP %r", res)
        result = res.get("result", {})
        return result.get("ad_info", {})

    async def location_lookup_by_coordinates(self, coordinates: str):
        params = dict(location=coordinates, key=self.api_key)
        res = await self.request("/ws/geocoder/v1", params=params)
        if res.get("status") != 0:
            self.logger.warning(
                "Failed to query location by coordinates %r", res
            )
        result = res.get("result", {})
        return result.get("ad_info", {})

    async def location_lookup_by_keyword(self, keyword: str):
        params = dict(keyword=keyword, key=self.api_key)
        res = await self.request("/ws/district/v1/search", params=params)
        if res.get("status") != 0:
            self.logger.warning("Failed to query location by keyword %r", res)
        results = res.get("result", [])
        if not results or not results[0]:
            return {}
        location = results[0][0]["location"]
        lat = location["lat"]
        lng = location["lng"]
        return await self.location_lookup_by_coordinates(f"{lat},{lng}")

    async def location_lookup(self, query: str):
        ad_info = None
        ip_candidates = re.findall(
            r"\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b", query
        )
        if ip_candidates:
            ad_info = await self.location_lookup_by_ip(ip_candidates[0])

        coordinates = re.findall(r"^(-?\d+\.\d+?),\s*(-?\d+\.\d+?)$", query)
        if coordinates:
            ad_info = await self.location_lookup_by_coordinates(
                ",".join(coordinates[0])
            )

        if not ad_info:
            ad_info = await self.location_lookup_by_keyword(query)

        return ad_info


async def query_current_weather(api_key: str, query: str):
    """
    To query the QQ (Tencent) Weather API for real-time weather data in a
    location of your choice.

    :param api_key: Tencent Map WebServiceAPI key.
    :param query: Pass a single location identifier to the API and
                  auto-detect the associated location. For Example:
        北京市 - Location Name
        110105 - adcode (行政区划代码)
        39.90469,116.40717 - Coordinates (Lat/Lon)
        61.135.17.68 - IP Address.
    :return: real-time weather data.
    """
    if not api_key:
        raise ValueError("Please provide tencent map api key")

    if not query:
        raise ValueError("Empty query")

    async with aiohttp.ClientSession() as session:
        qq_map = QQMap(api_key, session=session, logger=qq_logger)
        qq_weather = QQWeather(session=session, logger=qq_logger)
        ad_info = await qq_map.location_lookup(query)
        province, city = ad_info.get("province"), ad_info.get("city")

        res = await qq_weather.fetch_current_weather(province, city)
        res.update(location=ad_info)
        return res


async def query_weather_forecast(
    api_key: str, query: str, forecast_days: int = 7
):
    """
    The QQ (Tencent) Weather API is capable of returning weather forecast data
    for up to 7 days into the future.

    :param api_key: Tencent Map WebServiceAPI key.
    :param query: Pass a single location identifier to the API and
                  auto-detect the associated location. For Example:
        北京市 - Location Name
        110105 - adcode (行政区划代码)
        39.90469,116.40717 - Coordinates (Lat/Lon)
        61.135.17.68 - IP Address.
    :param forecast_days: The number of days for which the API returns forecast
                          data (Default: 7 days).
                          If pass forecast_days is 1, it will return weather
                          data split hourly.
    :return: forecast weather data.
    """
    if not api_key:
        raise ValueError("Please provide tencent map api key")

    if not query:
        raise ValueError("Empty query")

    if forecast_days > 7 or forecast_days < 0:
        raise ValueError("Invalid forecast days")

    async with aiohttp.ClientSession() as session:
        qq_map = QQMap(api_key, session=session, logger=qq_logger)
        qq_weather = QQWeather(session=session, logger=qq_logger)
        ad_info = await qq_map.location_lookup(query)
        province, city = ad_info.get("province"), ad_info.get("city")

        res = await qq_weather.fetch_weather_forecast(
            province, city, forecast_days
        )
        res.update(location=ad_info)
        return res
